prov/efa/src/rxr/rxr_read.c

```diff
@@ -230,12 +237,14 @@ int rxr_read_init_iov(struct rxr_ep *ep,
 	struct rxr_peer *peer;
 
 	peer = rxr_ep_get_peer(ep, tx_entry->addr);
-	if (peer->is_local) {
+	if (rxr_env.enable_shm_transfer && peer->is_local) {
+		if (!rxr_env.shm_cma_cap)
+			rxr_inline_mr_reg(ep, tx_entry);
 		for (i = 0; i < tx_entry->iov_count; ++i) {
-			assert(!tx_entry->mr[i]);
 			read_iov[i].addr = (uint64_t)tx_entry->iov[i].iov_base;
 			read_iov[i].len = tx_entry->iov[i].iov_len;
-			read_iov[i].key = 0;
```

Comment:
I do not think you need call rxr_inline_mr_reg() because this function also does memory registration will you can use.

I think you just need to adjust the condition a little bit to:

```
if (tx_entry->desc[I]) {
    // no registration 
} else if (shm && rxr_env.shm_cma_cap) {
    rma_iov[I].key = 0
} else {
    // memory registration and set rma_iov[I].key.
}
```

Reply:
Thank you. Updated.